doc: update life of a search query doc to be current (#29736)
sourcegraph/sourcegraph: doc/dev/background-information/architecture/life-of-a-search-query.md

## Frontend

> The frontend uses a [Streaming API](../../../api/stream_api/index.md). The frontend implements it [here](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+func+%28h+*streamHandler%29+ServeHTTP). The frontend implements the GraphQL search resolver [here](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+func+%28r+*schemaResolver%29+Search%28). The Streaming API is used by the browser. Historically we served results via GraphQL and there are still many clients who use this API. Internally Sourcegraph search is streaming based.

Reviewer: ```suggestion
The frontend uses a [Streaming API](../../../api/stream_api/index.md). The frontend implements it [here](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+func+%28h+*streamHandler%29+ServeHTTP). The Streaming API is used by the browser. Historically we served results via [GraphQL](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+func+%28r+*schemaResolver%29+Search%28) and there are still many clients who use this API. Internally Sourcegraph search is streaming based.
```

> First, the frontend takes the query and [creates a plan of jobs](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+func+%28r+*searchResolver%29+toSearchInputs&patternType=literal) to execute concurrently. A job is a specific query against a backend. For example [converting a Sourcegraph query into a Zoekt query](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+func+querytozoektquery&patternType=literal) for our indexed search backend. The comments in the job creation function are worth reading for more details.

Reviewer: ```suggestion
First, the frontend takes the query and [creates a plan of jobs](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+func+%28r+*searchResolver%29+toSearchInputs&patternType=literal) to execute concurrently. A job is a specific query against a backend. For example [here we convert a Sourcegraph query into a Zoekt query](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+func+querytozoektquery&patternType=literal) for our indexed search backend. The comments in the job creation function are worth reading for more details.
```

> Most Sourcegraph queries can be directly translated into Zoekt queries without consulting our database of repositories. However, not all repositories or revisions are indexed. So we need to work out what isn't in zoekt so we can do unindexed queries against Searcher. So the frontend [determines which repository@revision combinations are indexed by zoekt](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+zoektIndexedRepos%28+file:indexed_search.go&patternType=literal) by [consulting an in-memory cache](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+type+cachedSearcher). See [RepoSubsetTextSearch](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+type+RepoSubsetTextSearch&patternType=literal) for how searcher is queried.

Reviewer: ```suggestion
Most Sourcegraph queries can be directly translated into Zoekt queries without consulting our database of repositories. However, not all repositories or revisions are indexed. So we need to work out what isn't indexed by Zoekt so we can do unindexed queries against Searcher. So the frontend [determines which repository@revision combinations are indexed by Zoekt](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+zoektIndexedRepos%28+file:indexed_search.go&patternType=literal) by [consulting an in-memory cache](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+type+cachedSearcher). See [RepoSubsetTextSearch](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+type+RepoSubsetTextSearch&patternType=literal) for how searcher is queried.
```

To decide what to index [zoekt-sourcegraph-indexserver](https://sourcegraph.com/github.com/sourcegraph/zoekt/-/tree/cmd/zoekt-sourcegraph-indexserver) sends an [HTTP Get request to the frontend internal API](https://sourcegraph.com/search?q=context:global+r:github.com/sourcegraph/+-file:%28doc%7Ctest%7Cspec%29+%22/repos/index%22+fork:yes&patternType=regexp) at most once per minute to fetch the list of repository names to index. For each repository the indexserver will compare what Sourcegraph wants indexed (commit, configuration, etc.) to what is already indexed on disk and will start an index job for anything that is missing.

What we index in a repository is affected by admin configuration (branches, large file allow list). For each repository the indexserver [asks the frontend for configuration](https://sourcegraph.com/search?q=context:global+r:github.com/sourcegraph/+-file:%28doc%7Ctest%7Cspec%29+%22/search/configuration%22+fork:yes&patternType=regexp).

> It uses relies on [git shallow clones](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+-file:%28test%7Cspec%7Cdoc%29+GitUploadPack&patternType=literal) to fetch the contents of the branches to index. It then calls out to `zoekt-git-index` which creates 1 or more shards containing the indexes used by `zoekt-webserver`.

Reviewer: ```suggestion
It uses on [git shallow clones](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+-file:%28test%7Cspec%7Cdoc%29+GitUploadPack&patternType=literal) to fetch the contents of the branches to index. It then calls out to `zoekt-git-index` which creates 1 or more shards containing the indexes used by `zoekt-webserver`.
```

Reviewer: I would have expected this link for shallow clones. Wdyt? https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/zoekt%24+-file:_test+--depth%3D1&patternType=literal

Author: You are right, I'll update.
